File: 1001-1500/1481_findLeastNumOfUniqueInts.py
class Solution:
    def findLeastNumOfUniqueInts(self, arr: List[int], k: int) -> int:
        c = collections.Counter(arr)
        heap = []
        for key, val in c.items():
            heapq.heappush(heap, (val, key))
        while k:
            if not heap:
                return 0
            count, num = heapq.heappop(heap)
            if count > k:
                return len(heap) + 1
            elif count == k:
                return len(heap)
            else:
                k -= count
        return len(heap)

# Alternative solution

import heapq

class Solution:
    def findLeastNumOfUniqueInts(self, arr: List[int], k: int) -> int:
        freq = dict()
        for num in arr:
            freq[num] = freq.get(num, 0) + 1
        heap = []
        for key, val in freq.items():
            heapq.heappush(heap, (val, key))
        while k > 0 and heap:
            count, _ = heapq.heappop(heap)
            k -= count
        return len(heap) if k >= 0 else len(heap) + 1

File: 1001-1500/1481_findLeastNumOfUniqueInts.java
class Tuple implements Comparable<Tuple> {
    private int key;
    private int value;

    public Tuple(int key, int value) {
        this.key = key;
        this.value = value;
    }

    @Override
    public int compareTo(Tuple tuple) {
        return this.value - tuple.value;
    }

    public int getKey() {
        return this.key;
    }

    public int getValue() {
        return this.value;
    }
}

class Solution {
    public int findLeastNumOfUniqueInts(int[] arr, int k) {
        HashMap<Integer, Integer> counter = new HashMap<>();
        PriorityQueue<Tuple> pq = new PriorityQueue<>();
        int uniqueNumbers = 0;

        for (int num: arr) counter.put(num, counter.getOrDefault(num, 0) + 1);

        for (Map.Entry<Integer, Integer> entry: counter.entrySet()) {
            int key = entry.getKey();
            int value = entry.getValue();
            pq.add(new Tuple(key, value));
        }

        while (k >= 0 && !pq.isEmpty()) {
            Tuple tuple = pq.poll();
            if (k - tuple.getValue() >= 0) uniqueNumbers++;
            k -= tuple.getValue();
        }

        return counter.size() - uniqueNumbers;
    }
}
